source/slang/slang-check-overload.cpp

```diff
@@ -2385,6 +2390,35 @@ namespace Slang
         return argsListBuilder.produceString();
     }
 
+
+    /// We allow a special case for when `funcExpr` is expected to be a Default initializer
+    /// but none exist. Note, we cannot just create a default initializer for every variable
+    /// since then we are introducing initialization to every variable through an indirect
+    /// init returning data. Instead we will call `$DefaultInit` through this logic below.
+    ///     struct S { int x; }
+    ///     S s = S(); // This will call S.$DefaultInit().
+    Expr* _tryOverloadWithDefaultInit(SemanticsVisitor* visitor, SemanticsVisitor::OverloadResolveContext& context, Expr* expr, OverloadCandidate* bestCandidate)
```

Comment:
Can you explain this again? I still don't understand what it is trying to do. For the example shown in the comment here, haven't we already synthesized the default ctors for `S`, so that the overload resolution will resolve into a call to that defualt ctor already? Why do we need to run the coercion logic again here?

Reply:
This is code that mostly carries over from how TOT did things (put into its own function). I will break down why we have this function into 2 parts.

*What `_tryOverloadWithDefaultInit` does:*

If I type the following code, `_tryOverloadWithDefaultInit` will be used to resolve `MyStruct()`:
```
struct MyStruct
{
    float a;
};

[numthreads(1,1,1)]
void computeMain(int3 tid : SV_DispatchThreadID)
{
    MyStruct val = MyStruct(); // effectively `MyStruct val = {}`
}
```

*Why do we have this function:*

[using the code above] We have 2 requirements for any call to `MyStruct()` or related (Ex: `RayDesc()`):
1. We need to treat `MyStruct()` as a 'default-initialize'. This is a requirement since lots of code relies on this behavior.
2. We cannot treat `MyStruct()` as a 'default-initialize' using a compiler generated `init()` since this will force all uninitialized `struct` types into calling their `init()` ('default-initialize'). This is undesired behavior.

An example where rule 2 is important:
```
struct MyStruct
{
    float a;
};

[numthreads(1,1,1)]
void computeMain(int3 tid : SV_DispatchThreadID)
{
    MyStruct val1 = MyStruct(); // effectively `MyStruct val = {};`. This variable is default initialized.
    MyStruct val2; // Effectively `MyStruct val2;`. This variable is uninitialized.
}
```

Reply:
Create a sub-task to remove this hack: https://github.com/shader-slang/slang/issues/5138